File: SCT/branches/SCT_1.1.x_BRANCH/com.yakindu.statechart.model/plugins/org.mda4e.statemachine.model/src-gen/statemachine/Event.java
/**
 * Copyright (c) 2006-2009 committers of mda4e and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *     committers of mda4e (http://www.mda4e.org/) - initial API and implementation
 *
 */
package statemachine;


/**
 * <!-- begin-user-doc -->
 * A representation of the model object '<em><b>Event</b></em>'.
 * <!-- end-user-doc -->
 *
 * <p>
 * The following features are supported:
 * <ul>
 *   <li>{@link statemachine.Event#getTrigger <em>Trigger</em>}</li>
 * </ul>
 * </p>
 *
 * @see statemachine.StatemachinePackage#getEvent()
 * @model
 * @generated
 */
public interface Event extends DataElement {
	/**
	 * Returns the value of the '<em><b>Trigger</b></em>' attribute.
	 * The literals are from the enumeration {@link statemachine.TriggerTypes}.
	 * <!-- begin-user-doc -->
	 * <p>
	 * If the meaning of the '<em>Trigger</em>' attribute isn't clear,
	 * there really should be more of a description here...
	 * </p>
	 * <!-- end-user-doc -->
	 * @return the value of the '<em>Trigger</em>' attribute.
	 * @see statemachine.TriggerTypes
	 * @see #setTrigger(TriggerTypes)
	 * @see statemachine.StatemachinePackage#getEvent_Trigger()
	 * @model
	 * @generated
	 */
	TriggerTypes getTrigger();

	/**
	 * Sets the value of the '{@link statemachine.Event#getTrigger <em>Trigger</em>}' attribute.
	 * <!-- begin-user-doc -->
	 * <!-- end-user-doc -->
	 * @param value the new value of the '<em>Trigger</em>' attribute.
	 * @see statemachine.TriggerTypes
	 * @see #getTrigger()
	 * @generated
	 */
	void setTrigger(TriggerTypes value);

} // Event

File: SCT/tags/SCT_1.1.0_RELEASE/com.yakindu.statechart.codegenerator/examples/com.yakindu.statechart.codegenerator.examples.c/environment/event.cpp
/**
 * Copyright (c) 2010 committers of YAKINDU and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *     committers of YAKINDU - initial API and implementation
 */
#include "event.h"

Event::Event()
{
  EventPtr = new DummyEventHandle;
  ref_counter = new (uint32);
  (*ref_counter) = 1;
}

Event::Event(const Event& _E)
{
  EventPtr = _E.EventPtr;
  ref_counter = _E.ref_counter;
  (*ref_counter)++;
}

/*
Event::Event(const int64 to, bool active)
{
  EventPtr = new EventHandler(to, task, ID, active);
  ref_counter = new (uint32);
  (*ref_counter) = 1;
}
*/

Event::Event(EventHandle* ePtr)
{
  EventPtr = ePtr;
  ref_counter = new (uint32);
  (*ref_counter) = 1; // hope this is the first one
}

Event::~Event()
{
  (*ref_counter)--;
  if (!(*ref_counter)) {
    delete EventPtr;
    delete ref_counter;
  }
}

Event Event::clone()
{
  Event cloneEvent(EventPtr->clone());
  return cloneEvent;
}

EventHandle* Event::ptr()
{
  return (EventPtr);
}

void Event::setTimeout(int64 to)
{
  EventPtr->active = true;
  EventPtr->timeout = to;
}

int64 Event::getTimeout() const
{
  return(EventPtr->timeout);
}

void Event::setInactive()
{
  EventPtr->active = false;
}

bool Event::isActive() const
{
  return(EventPtr->active);
}

Event& Event::operator=(const Event& _E)
{
  if (this != &_E) {
    (*ref_counter)--;
    if (!(*ref_counter)) {
      delete EventPtr;
      delete ref_counter;
    }

    EventPtr = _E.EventPtr;
    ref_counter = _E.ref_counter;
    (*ref_counter)++;
  }

  return (*this);
}

bool Event::isExpired(int64 tics) const
{
  return (EventPtr->timeout <= tics);
}

bool Event::operator==(const Event& _E) const
{
  return (EventPtr == _E.EventPtr);
}

void Event::runTask()
{
  EventPtr->runTask();
}

Task& Event::getTask() const
{
  return(EventPtr->getTask());
}

bool Event::empty()
{
  return(&EventPtr->getTask() == 0);
}

bool Event::operator<(const Event& _E) const
{
  if (EventPtr->timeout >= _E.getTimeout())
    return (true);
  return (false);
}

bool isEventEqual::operator()(Event a, Event b) const
{
  if (a.getTimeout() == b.getTimeout())
    return (true);
  return (false);
}

bool isEventLess::operator()(Event a, Event b) const
{
  if (a.getTimeout() > b.getTimeout())
    return (true);
  return (false);
}
